### PR TITLE
Improve chances of DirectColor terminal detection

### PR DESCRIPTION
**Pull request by grendello**
While ncurses 6.1 adds the `RGB` capability to terminfo, the flag is not
commonly used in terminfo entries as of now. The `COLORTERM` environment
variable is, otoh, commonly present and set to either `truecolor` or
`24bit` value by terminal emulators which support direct color.

Add a check for this if `RGB` is false (or absent)

With this check in I was able to properly appreciate the awesomeness of
notcurses demos :)

### CODE REVIEW

**Comment by dankamongmen** [comment]
Heh. I'm so strongly tempted. In my initial preview email to @ThomasDickey (sent 2019-11-23 21:10:19 EDT, "sneak preview of notcurses", I wrote:

> It uses terminfo extensively, because I'm a sane individual. No
COLORTERM nonsense for this guy.

So I'm kinda torn here. On one side, yes, `COLORTERM` is the way in which certain terminals are indicating RGB support. On the other side, this gives me no portable terminfo-based way to emit the proper escapes. On the original side, terminfo as bundled with 6.1 doesn't give me a pure RGB `setaf`/`setab` either, IIRC (hence hand-built escapes in `term_esc_rgb()`, which I'd like to remove). So....

When I consider this kind of thing, I ask myself "what would Thomas E. Dickey do?" He's curmudgeonly and cranky, but by God the man has maintained ncurses for 25 years of excellence. Conservative excellence, but excellence. And he very clearly dislikes this Slang-originated methodology.

The "correct" answer, I suspect, is that you ought use a local terminfo override (e.g. `$HOME/.terminfo/blahblah` or the value of the `TERMINFO` environment variable; see `terminfo(5)`), and add the `rgb` capability. Then you lobby your upstream to add it to their terminal capabilities definition file in terminfo so that you don't have to carry this around for unto seventy times seven generations. But I understand that this is slightly esoteric for even technical users such as yourself.

I'm about to go play trivia. When I get back, I'll examine the `setaf` and `setab` escapes for various DirectColor-capable terminfo definitions. If they're RGB-capable and not doing and weird builtin downsampling, I'll switch to them, and be very hesistant about merging this. If they're not, and I have to do my own escapes anyway, I'll probably hold my nose and merge this.

One question: what was your TERM variable? For xterm, for instance, only `xterm-direct` gets DirectColor; `xterm-256color` wants an 8-bit indexed palette. Likewise for VTE, kitty, and alacritty. In my bashrc, I have:

```
if [ "$TERM" != "dumb" ] ; then
  # alacritty-direct is selected in alacritty.yml
  if [ "$TERM" == "xterm" ] ; then
    export TERM=xterm-direct
  elif [ "$TERM" == "xterm-kitty" ] ; then
    export TERM=kitty-direct
  fi
```

tl;dr: I'm aware of `COLORTERM`, and consciously refused to honor it. I'd really like to see the terminfo definitions properly updated.

**Comment by dankamongmen** [comment]
Alrightie @grendello , here's what I'm thinking. I took a look into the `xterm-direct` terminfo file shipped with Debian Unstable (6.1+20191019, the most recent release AFAIK). The `setaf` capability there is:

```
[schwarzgerat](0) $ infocmp | grep setaf
	setaf=\E[%?%p1%{8}%<%t3%p1%d%e38\:2\:\:%p1%{65536}%/%d\:%p1%{256}%/%{255}%&%d\:%p1%{255}%&%d%;m,
[schwarzgerat](0) $ 
```

So this looks like it passes through low colors. I can understand the reasoning for that, but it's definitely not what I want. I hacked things up to use `tputs(tiperm` with these strings, and indeed, sufficiently low colors get (wildly non-linearly) mapped to the system palette. These defects are particularly noticeable on the `view` and `maxcolors` demos. In addition, we send ~3-4% more bytes as a side-effect.

If this was because xterm wasn't properly functioning when sent these low colors as RGB values, I'd understand, but as it is, our hand-rolled setaf/setab for RGB seem to work fine in my testing (I'm sure there are situations where they don't, but I can't find any). If this is just for *application* backwards compatibility, screw it; we're a new library.

Since I'm going to stick with hand-consed RGB setaf/setab, might as well go whole-hog and say that "The environment variable COLORTERM indicates a capability of sending the escapes 48 and 38 together with a tripartite RGB (`0 <= c <= 255` for all three components) to specify fore- and background colors," with the intent of removing this when/if terminfo exports suitable pure RGB capabilities.

So, I'd like you to add a note to that effect in the README.md, at which point I'll merge this change. If you don't feel comfortable doing so, say as much, and I'll merge this and then write one myself. Thanks for the patch!

(ps one nit: I've been doing `type* ident` in this project, not `type *ident`)

File: src/lib/notcurses.c [hunk 1]
```diff
@@ -575,6 +575,10 @@ interrogate_terminfo(notcurses* nc, const notcurses_options* opts){
   char* longname_term = longname();
   fprintf(stderr, "Term: %s\n", longname_term ? longname_term : "?");
   nc->RGBflag = tigetflag("RGB") == 1;
+  if (nc->RGBflag == 0) {
+    const char *cterm = getenv("COLORTERM");
```
Comment (dankamongmen) [inline src/lib/notcurses.c]: nit: swap the * and the space. at minimum, a comment here would be nice.

is COLORTERM defined to anything besides "truecolor"/"24bit" that we might want to be aware of? we do quantization down to the palette if DirectColor isn't available.

**Comment by grendello** [comment]
> Heh. I'm so strongly tempted. In my initial preview email to @ThomasDickey (sent 2019-11-23 21:10:19 EDT, "sneak preview of notcurses"), I wrote:
> 
> > It uses terminfo extensively, because I'm a sane individual. No
> > COLORTERM nonsense for this guy.
> 
> So I'm kinda torn here. On one side, yes, `COLORTERM` is the way in which certain terminals are indicating RGB support. On the other side, this gives me no portable terminfo-based way to emit the proper escapes. On the original side, terminfo as bundled with 6.1 doesn't give me a pure RGB `setaf`/`setab` either, IIRC (hence hand-built escapes in `term_esc_rgb()`, which I'd like to remove). So....

I principle I agree with Thomas's sentiment, however... The problem here is that until very recently it's been impossible to indicate via terminfo whether or not a terminal supports direct color and `COLORTERM` (imperfect as it is) was found to be the common way to signal to the app that direct color is available. So it's ideal vs practicality (and practice) here. By checking only the terminfo database, there's no easy way for anyone to take advantage of DirectColor, not without hacking around. Hence this PR :)

> 
> When I consider this kind of thing, I ask myself "what would Thomas E. Dickey do?" He's curmudgeonly and cranky, but by God the man has maintained ncurses for 25 years of excellence. Conservative excellence, but excellence. And he very clearly dislikes this Slang-originated methodology.

It's a "to have or to have not" kind of choice as we are today, though :)

> 
> The "correct" answer, I suspect, is that you ought use a local terminfo override (e.g. `$HOME/.terminfo/blahblah` or the value of the `TERMINFO` environment variable; see `terminfo(5)`), and add the `rgb` capability. Then you lobby your upstream to add it to their terminal capabilities definition file in terminfo so that you don't have to carry this around for unto seventy times seven generations. But I understand that this is slightly esoteric for even technical users such as yourself.

It's not that complicated but it's impractical. Say you ship an app that takes full advantage of DirectColor and you want to do the Right Thing™, so what do you need to do? Easy, detect the terminal type (*not* by looking at the `TERM` variable, though) and create the required entry, then probably restart the app. Sounds simple but it's anything but - the detection is possible but hacky at best, not to mention that I don't think apps should modify user config that doesn't pertain to their app directly this way.

> 
> I'm about to go play trivia. When I get back, I'll examine the `setaf` and `setab` escapes for various DirectColor-capable terminfo definitions. If they're RGB-capable and not doing and weird builtin downsampling, I'll switch to them, and be very hesistant about merging this. If they're not, and I have to do my own escapes anyway, I'll probably hold my nose and merge this.

Sounds good :)

> 
> One question: what was your TERM variable? For xterm, for instance, only `xterm-direct` gets DirectColor; `xterm-256color` wants an 8-bit indexed palette. Likewise for VTE, kitty, and alacritty. In my bashrc, I have:

In my case it was `xterm-kitty`, and nope, it doesn't have the RGB flag in Ubuntu 19.10 :(

> 
> ```
> if [ "$TERM" != "dumb" ] ; then
>   # alacritty-direct is selected in alacritty.yml
>   if [ "$TERM" == "xterm" ] ; then
>     export TERM=xterm-direct
>   elif [ "$TERM" == "xterm-kitty" ] ; then
>     export TERM=kitty-direct
>   fi
> ```
> 
> tl;dr: I'm aware of `COLORTERM`, and consciously refused to honor it. I'd really like to see the terminfo definitions properly updated.

I actually agree with you, however the reality doesn't match the ideal today :(

**Comment by grendello** [comment]
> Alrightie @grendello , here's what I'm thinking. I took a look into the `xterm-direct` terminfo file shipped with Debian Unstable (6.1+20191019, the most recent release AFAIK). The `setaf` capability there is:
> 
> ```
> [schwarzgerat](0) $ infocmp | grep setaf
> 	setaf=\E[%?%p1%{8}%<%t3%p1%d%e38\:2\:\:%p1%{65536}%/%d\:%p1%{256}%/%{255}%&%d\:%p1%{255}%&%d%;m,
> [schwarzgerat](0) $ 
> ```
> 
> So this looks like it passes through low colors. I can understand the reasoning for that, but it's definitely not what I want. I hacked things up to use `tputs(tiperm` with these strings, and indeed, sufficiently low colors get (wildly non-linearly) mapped to the system palette. These defects are particularly noticeable on the `view` and `maxcolors` demos. In addition, we send ~3-4% more bytes as a side-effect.
> 

There's also this way to detect the capability which might be more up your alley, but I think it's slightly overcomplicated for such a simple purpose: https://gist.github.com/XVilka/8346728#querying-the-terminal

> If this was because xterm wasn't properly functioning when sent these low colors as RGB values, I'd understand, but as it is, our hand-rolled setaf/setab for RGB seem to work fine in my testing (I'm sure there are situations where they don't, but I can't find any). If this is just for _application_ backwards compatibility, screw it; we're a new library.
> 
> Since I'm going to stick with hand-consed RGB setaf/setab, might as well go whole-hog and say that "The environment variable COLORTERM indicates a capability of sending the escapes 48 and 38 together with a tripartite RGB (`0 <= c <= 255` for all three components) to specify fore- and background colors," with the intent of removing this when/if terminfo exports suitable pure RGB capabilities.
> 
> So, I'd like you to add a note to that effect in the README.md, at which point I'll merge this change. If you don't feel comfortable doing so, say as much, and I'll merge this and then write one myself. Thanks for the patch!

Will do! :)

> 
> (ps one nit: I've been doing `type* ident` in this project, not `type *ident`)

Muscle memory, will fix :)

**Comment by dankamongmen** [comment]
@grendello , thanks for becoming the first contributor (besides me) to notcurses, and what was an uncommonly pleasant and reasonable exchange for the first incoming PR. Merging.

**Comment by grendello** [comment]
> @grendello , thanks for becoming the first contributor (besides me) to notcurses, and what was an uncommonly pleasant and reasonable exchange for the first incoming PR. Merging.

Pleasure to contribute :) Keep doing the good work and thanks for taking time to create this library !
